test/units/plugins/inventory/test_constructed.py

```diff
@@ -102,3 +103,67 @@ def test_keyed_parent_groups(inventory_module):
     all_regions = inventory_module.inventory.groups['region_list']
     assert all_regions.child_groups == [region_group]
     assert region_group.hosts == [host1, host2]
+
+
+def test_parent_group_templating(inventory_module):
+    inventory_module.inventory.add_host('cow')
+    inventory_module.inventory.set_variable('cow', 'sound', 'mmmmmmmmmm')
+    inventory_module.inventory.set_variable('cow', 'nickname', 'betsy')
+    host = inventory_module.inventory.get_host('cow')
+    keyed_groups = [
+        {
+            'key': 'sound',
+            'prefix': 'sound',
+            'parent_group': '{{ nickname }}'
+        },
+        {
+            'key': 'nickname',
+            'prefix': '',
+            'separator': '',
+            'parent_group': 'nickname'  # statically-named parent group, conflicting with hostvar
+        },
+        {
+            'key': 'nickname',
+            'separator': '',
+            'parent_group': '{{ location | default("field") }}'
+        }
+    ]
+    inventory_module._add_host_to_keyed_groups(
+        keyed_groups, host.vars, host.name, strict=True
+    )
+    # first keyed group, "betsy" is a parent group name dynamically generated
+    betsys_group = inventory_module.inventory.groups['betsy']
+    assert [child.name for child in betsys_group.child_groups] == ['sound_mmmmmmmmmm']
+    # second keyed group, "nickname" is a statically-named root group
+    nicknames_group = inventory_module.inventory.groups['nickname']
+    assert [child.name for child in nicknames_group.child_groups] == ['betsy']
+    # second keyed group actually generated the parent group of the first keyed group
+    # assert that these are, in fact, the same object
+    assert nicknames_group.child_groups[0] == betsys_group
+    # second keyed group has two parents
+    locations_group = inventory_module.inventory.groups['field']
+    assert [child.name for child in locations_group.child_groups] == ['betsy']
+
+
+def test_parent_group_templating_error(inventory_module):
+    inventory_module.inventory.add_host('cow')
+    inventory_module.inventory.set_variable('cow', 'nickname', 'betsy')
+    host = inventory_module.inventory.get_host('cow')
+    keyed_groups = [
+        {
+            'key': 'nickname',
+            'separator': '',
+            'parent_group': '{{ location.barn-yard }}'
+        }
+    ]
+    with pytest.raises(AnsibleParserError) as err_message:
+        inventory_module._add_host_to_keyed_groups(
+            keyed_groups, host.vars, host.name, strict=True
+        )
+        assert 'Could not generate parent group' in err_message
+    inventory_module._add_host_to_keyed_groups(
+        keyed_groups, host.vars, host.name, strict=False
+    )
```

Comment:
Given the location where the exception is raised, I don't think you should need to call `_add_host_to_keyed_groups` a second time.

In the event of the template error, the group `betsy` should still be created, but without any parent groups, because `{{ location.barn-yard }}` couldn't be evaluated. I think this behavior is ideal.

Sometimes, a key will only be present for some hosts, but not others. In that case, you may still want to apply the first level of grouping, but no apply the grandparent group.

Reply:
I'm not really sure what the harm is in checking that strict=False is ignored.

Reply:
> I'm not really sure what the harm is in checking that strict=False allows parent groups that can't be templated to be ignored.

Yes, I agree, what I'm saying is that L164-L166 is not doing this.

I would say that you could do that as a separate test without loss of meaning, but other things are broken... `inventory_module` really shouldn't be module scope, since multiple tests are using the same names. If that scope was taken away, then you could meaningfully test the strict=False case in another isolated test.

I'm just saying that the way you've written this test, running with strict=False creates the same stuff as running with strict=True. It's just distracting that an exception is being caught. Just because the exception was caught doesn't mean the inventory content wasn't created.

Reply:
I also don't think you necessary need to test that separately, just deleting those lines would effectively test it, but wouldn't have this confusion.

Reply:
I'm not sure i agree, if the definition of the compose clause is wrong, all the actions of that clause should be skipped.

Reply:
Hm... Ok, I've moved it so the failure happens before adding any groups and test that the group is never added.

Reply:
> if true the whole plugin output should be reverted

I was aware of a RFE to have the inventory loader revert its work for a particular source in the list if an error occurs. The last I knew of this, it was not yet a thing.

If this was a thing, then I would expect an assertion after the try-except block which establishes that the "betsy" group does not exist.

For these tests, it's easier to avoid the question of the existence or non-existence of that feature, either by testing only 1 scenario, or breaking it out into 2 mutually isolated tests.

Reply:
the global revert is a bug, which we have to fix, but the general behaviour should always be the same, if you have an incorrect definition, fail the whole grouping, don't do partials ... strict=false allows for other definitions to continue but the user must explicitly ask for that.